### ISSUE TITLE
Magento 2 has no ability to order an event's observers execution

### ISSUE THREAD

**Issue by mage2pro** [magento/magento2#2354]
https://mage2.pro/t/200

For example, if there are mutiple observers for the `catalog_block_product_status_display` event then the Magento 2 behavior becomes unpredictable because we can not set the observer's ordering.

**Comment by rhamoudi** [comment]
Have you tested that the ordering isn't done by reading the xml from top to bottom?

**Comment by Yonn-Trimoreau** [comment]
+1

It will be really useful for ordering menu items when observing on `page_block_html_topmenu_gethtml_before` event.

**Comment by ntzz** [comment]
@Yonn-Trimoreau Is possible to order top menu yet?

**Comment by Yonn-Trimoreau** [comment]
Yes, you can plug an observer to the event "page_block_html_topmenu_gethtml_before" and recreate your own nodes. That's the only way I found.

In file Vendor/ModuleName/etc/frontend/events.xml :

```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
    <event name="page_block_html_topmenu_gethtml_before">
        <observer name="add_items_to_menu" instance="Vendor\ModuleName\Observer\AddItemsToMenu"/>
    </event>
</config>
```

In file Vendor/ModuleName/Observer/AddItemsToMenu.php :

```
<?php
namespace Vendor\ModuleName\Observer;

use Magento\Framework\Event\Observer;
use Magento\Framework\Event\ObserverInterface;

class AddItemsToMenu implements ObserverInterface
{
    /**
     * @var \Magento\Store\Model\StoreManagerInterface
     */
    protected $_storeManager;

    /**
     * @var \Magento\Framework\UrlInterface
     */
    protected $_urlInterface;

    /**
     * AddItemsToMenu Observer constructor.
     * @param \Magento\Store\Model\StoreManagerInterface $storeManager
     * @param \Magento\Framework\UrlInterface $urlInterface
     */
    public function __construct(
        \Magento\Store\Model\StoreManagerInterface $storeManager,
        \Magento\Framework\UrlInterface $urlInterface
    ) {
        $this->_storeManager = $storeManager;
        $this->_urlInterface = $urlInterface;
    }

    /**
     * @param Observer $observer
     * @return void
     */
    public function execute(\Magento\Framework\Event\Observer $observer)
    {
        /** @var \Magento\Framework\Data\Tree\Node $parentCategoryNode */
        $parentCategoryNode = $observer->getMenu();
        $block = $observer->getEvent()->getBlock();
        $tree = $parentCategoryNode->getTree();
        $currentUrl = $this->_urlInterface->getCurrentUrl();

        // Create menu array
        $nodesData = array(
            array(
                "name" => "Home",
                "id" => "home",
                "url" => $url = $block->getUrl(''),
                "has_active" => false,
                "is_active" => $currentUrl == $url ? true : false,
                "extra_class" => "home",
                "order" => 1,
            ),
            array(
                "name" => "Menu1",
                "id" => "menu1",
                "url" => $url = $block->getUrl('url/menu1'),
                "has_active" => false,
                "is_active" => $currentUrl == $url ? true : false,
                "extra_class" => "menu1",
                "order" => 2,
            ),
            [...]
        );

        foreach ($nodesData as $nodeData) {
            $categoryNode = new \Magento\Framework\Data\Tree\Node($nodeData, 'id', $tree, $parentCategoryNode);
            $parentCategoryNode->addChild($categoryNode);
        }
    }
}
```

**Comment by Yonn-Trimoreau** [comment]
In file Vendor/ModuleName/etc/frontend/events.xml :
```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
    <event name="page_block_html_topmenu_gethtml_before">
        <observer name="add_items_to_menu" instance="Vendor\ModuleName\Observer\AddItemsToMenu"/>
    </event>
</config>
```

In file Vendor/ModuleName/Observer/AddItemsToMenu.php :
```
<?php
namespace Vendor\ModuleName\Observer;

use Magento\Framework\Event\Observer;
use Magento\Framework\Event\ObserverInterface;

class AddItemsToMenu implements ObserverInterface
{
    /**
     * @var \Magento\Store\Model\StoreManagerInterface
     */
    protected $_storeManager;

    /**
     * @var \Magento\Framework\UrlInterface
     */
    protected $_urlInterface;

    /**
     * AddItemsToMenu Observer constructor.
     * @param \Magento\Store\Model\StoreManagerInterface $storeManager
     * @param \Magento\Framework\UrlInterface $urlInterface
     */
    public function __construct(
        \Magento\Store\Model\StoreManagerInterface $storeManager,
        \Magento\Framework\UrlInterface $urlInterface
    ) {
        $this->_storeManager = $storeManager;
        $this->_urlInterface = $urlInterface;
    }

    /**
     * @param Observer $observer
     * @return void
     */
    public function execute(\Magento\Framework\Event\Observer $observer)
    {
        /** @var \Magento\Framework\Data\Tree\Node $parentCategoryNode */
        $parentCategoryNode = $observer->getMenu();
        $block = $observer->getEvent()->getBlock();
        $tree = $parentCategoryNode->getTree();
        $currentUrl = $this->_urlInterface->getCurrentUrl();

        // Create menu array
        $nodesData = array(
            array(
                "name" => "Home",
                "id" => "home",
                "url" => $url = $block->getUrl(''),
                "has_active" => false,
                "is_active" => $currentUrl == $url ? true : false,
                "extra_class" => "home",
                "order" => 1,
            ),
            array(
                "name" => "Menu1",
                "id" => "menu1",
                "url" => $url = $block->getUrl('url/menu1'),
                "has_active" => false,
                "is_active" => $currentUrl == $url ? true : false,
                "extra_class" => "menu1",
                "order" => 2,
            ),
            [...]
        );

        foreach ($nodesData as $nodeData) {
            $categoryNode = new \Magento\Framework\Data\Tree\Node($nodeData, 'id', $tree, $parentCategoryNode);
            $parentCategoryNode->addChild($categoryNode);
        }
    }
}
```

**Comment by BarryCarlyon** [comment]
As far as I can tell, this will only add items onto the end of the menu.

I am as yet to find a way to add items to the start of the menu.

(But using the same "name":

```
<observer name="catalog_add_topmenu_items" instance="Magento\Catalog\Observer\AddCatalogToTopmenuItemsObserver" />
```

so

```
<observer name="catalog_add_topmenu_items" instance="Vendor\ModuleName\Observer\AddItemsToMenu"/
```

Instead 

Means you can override it it seems)

**Comment by BarryCarlyon** [comment]
But really an

```
$parentCategoryNode->prependChild($categoryNode);
```

Would be more useful

**Comment by BarryCarlyon** [comment]
I've ended up with:

> etc/frontend/events.xml

```
<?xml version="1.0"?>
<!--
/**
 * Copyright © 2016 Magento. All rights reserved.
 * See COPYING.txt for license details.
 */
-->
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
    <event name="page_block_html_topmenu_gethtml_before">
        <observer name="catalog_add_topmenu_items" instance="Vendor\Module\Observer\TopMenu" />
    </event>
</config>
```

> Observer/TopMenu.php

```
<?php

namespace Vendor\Module\Observer;

use Magento\Framework\Event\ObserverInterface;
use Magento\Framework\Data\Tree\Node;

class Topmenu implements ObserverInterface {
    /**
     * @var \Psr\Log\LoggerInterface
     */
    protected $logger;

    /**
     * @param \Magento\Catalog\Observer\AddCatalogToTopmenuItemsObserver $orig
     * @param \Psr\Log\LoggerInterface $logger
     */
    public function __construct(
        \Magento\Catalog\Observer\AddCatalogToTopmenuItemsObserver $orig,
        \Psr\Log\LoggerInterface $logger
    ) {
        $this->logger = $logger;

        $this->orig = $orig;
    }

    /**
     * Apply customized static files to frontend
     *
     * @param \Magento\Framework\Event\Observer $observer
     * @return void
     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
     */
    public function execute(\Magento\Framework\Event\Observer $observer)
    {
        $menu = $observer->getMenu();
        $tree = $menu->getTree();

        $data = [
            'name'      => 'TestING',
            'id'        => 'test',
            'url'       => '/test/',
            'is_active' => false
        ];

        $node = new Node($data, 'id', $tree, $menu);
        $menu->addChild($node);

        // add cats by calling the original observer
        $this->orig->execute($observer);
    }
}
```

**Comment by Yonn-Trimoreau** [comment]
My solution rebuilds the tree entirely, it does not only add nodes to the end ;)

**Comment by piotrekkaminski** [comment]
Thank you for your submission.

We recently made some changes to the way we process GitHub submissions to more quickly identify and respond to core code issues.

Feature Requests and Improvements should now be submitted to the new [Magento 2 Feature Requests and Improvements](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) forum (see details [here](https://community.magento.com/t5/News-Announcements/Improvements-to-GitHub-Management/m-p/44572#M96)).

We are closing this GitHub ticket and have moved your request to the new [forum](https://community.magento.com/t5/Magento-2-Feature-Requests-and/Magento-2-has-no-ability-to-order-an-event-s-observers-execution/idi-p/45359).

**Comment by vincentwinkel** [comment]
Ok, you moved to forums, but where is the solution ?
I'm also facing this issue. I added an event to block some orders after save (with personal logics) and the payment observer always update them to processing state.

Same problem with quotes. For this one, I found a workaround creating a plugin aroundExecute() on the observer. A plugin on a observer... what's the next step ? Overriding xslt config to inject overridden dependencies into plugins to rewrite observers in simply order to display "hello customer" on the front? Maybe we don't have the same definition of "modularity".
